daffodil-runtime1/src/main/scala/org/apache/daffodil/layers/ByteSwapTransformer.scala

```diff
@@ -0,0 +1,184 @@
+        case Filling => {
+          c = jis.read()
+          if (c == -1) {
+            state = Draining
+          } else {
+            stack.push(c)
+            if (stack.size() == wordsize) {
+              state = Emptying
+            }
+          }
+        }
+        case Emptying => {
+          if (stack.isEmpty()) {
+            state = Filling
+          } else {
+            c = stack.pop()
+            return c
+          }
+        }
+        case Draining => {
+          if (stack.isEmpty()) {
+            state = Done
+            return -1
+          } else {
+            c = stack.pop()
+            return c
+          }
+        }
+        case Done =>
+          Assert.invariantFailed("Done state not allowed.")
+      }
+    }
+    Assert.invariantFailed("No fall through to here.")
+  }
+}
+
+class ByteSwapOutputStream(wordsize: Int, jos: OutputStream)
+  extends OutputStream {
+
+  private val stack: Deque[Byte] = new ArrayDeque[Byte](wordsize)
+  private var closed = false
+
+  override def close(): Unit = {
+    if (!closed) {
+      while (!stack.isEmpty()) {
+        jos.write(stack.pop())
+      }
+      jos.close()
+      closed = true
+    }
+  }
+
+  override def write(bInt: Int): Unit = {
+    Assert.usage(!closed)
+    stack.push(bInt.toByte)
+    if (stack.size() == wordsize) {
+      while(!stack.isEmpty()) {
+        jos.write(stack.pop())
+      }
+    }
+  }
+}
+
+class ByteSwapTransformer(wordsize: Int, layerLengthInBytesEv: LayerLengthInBytesEv)
+  extends LayerTransformer() {
+
+  override def wrapLayerDecoder(jis: java.io.InputStream) = {
+    val s = new ByteSwapInputStream(wordsize, jis)
+    s
+  }
+
+  override def wrapLimitingStream(jis: java.io.InputStream, state: PState) = {
+    val layerLengthInBytes: Int = layerLengthInBytesEv.evaluate(state).toInt
+
+    val s = new ExplicitLengthLimitingStream(jis, layerLengthInBytes)
+    s
+  }
+
+  override protected def wrapLayerEncoder(jos: java.io.OutputStream): java.io.OutputStream = {
+    val s = new ByteSwapOutputStream(wordsize, jos)
+    s
+  }
+
+  override protected def wrapLimitingStream(jos: java.io.OutputStream, state: UState): java.io.OutputStream = {
+    jos // just return jos. The way the length will be used/stored is by way of
+    // taking the content length of the enclosing element. That will measure the
+    // length relative to the "ultimate" data output stream.
+  }
+}
+
+sealed abstract class ByteSwapTransformerFactory(wordsize: Int, name: String)
+  extends LayerTransformerFactory(name) {
+
+  override def newInstance(maybeLayerCharsetEv: Maybe[LayerCharsetEv],
+    maybeLayerLengthKind: Maybe[LayerLengthKind],
+    maybeLayerLengthInBytesEv: Maybe[LayerLengthInBytesEv],
+    maybeLayerLengthUnits: Maybe[LayerLengthUnits],
+    maybeLayerBoundaryMarkEv: Maybe[LayerBoundaryMarkEv],
+    trd: TermRuntimeData): LayerTransformer = {
+
+    val xformer = new ByteSwapTransformer(wordsize, maybeLayerLengthInBytesEv.get)
+    xformer
```

Comment:
It looks like this only supports layerLengthKind="explicit". Does it make sense to also implement implicit or boundarymark? If not, do we need to have schema definition checks to ensure the appropriate layer properties are provided? I don't see the checks for other transformers, so maybe we're missing those as well?

Reply:
I think for many layer transforms only certain layerLengthKinds make sense. So code should reject the other layerLengthKinds as incompatible.

Reply:
Many layer transforms only 1 or a subset of the layer lengthKinds make any sense. See our limited doc on layers on the wiki here: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=75979671

(which would be good to update also once this change is merged.)